### PR TITLE
Expand FilterQuality API docs

### PR DESCRIPTION
**Pull request by yjbanov**
Document more precisely the effects of the `FilterQuality` enum values with a focus on relative quality when scaling images _up_ vs scaling them _down_. It turns out that the algorithm used for `high` is frequently _worse_ than `medium` when scaling down. Only when scaling images up do you get increase in quality as you go none => low => medium => high.

This PR depends on https://github.com/flutter/assets-for-api-docs/pull/134.

Fixes https://github.com/flutter/flutter/issues/76187

### CODE REVIEW

**Comment by flar** [comment]
I feel like these clarifications are more of a report card for the implementation rather than documentation of the intention. The way they are phrased could end up cementing the expectation of the results in lieu of fixing the underlying problems.

**Comment by yjbanov** [comment]
@flar

> I feel like these clarifications are more of a report card for the implementation rather than documentation of the intention.

I think it's still better to have them than not, but...

> in lieu of fixing the underlying problems.

...I agree that we should fix it eventually. Based on the cases I've seen so far, I would just remove `low` altogether and remap all options as follows:

| Value | Old semantic | Action | New semantic |
| --- | --- | --- | --- |
| none | low quality | no change | low quality |
| medium | smooth | remapped to what `low` does today | medium quality, default in framework |
| high | who knows? | remapped to what `medium` does today | high quality, opt-in |
| ~~low~~ | ~~smoothish~~ | ~~deprecate it~~ | ~~gone~~ |

Implicitly, the current algorithm used by `high` is gone too. I haven't seen any examples where it would produce a better result compared to `medium`.

For advanced use-cases the Skia team tells me we can provide an explicit sampler. This way users can have a high degree of control. That would require an extension in our `dart:ui` API.

It would be a breaking change, so I wouldn't rush just yet.

**Comment by flar** [comment]
We have the following possible techniques that can be applied at a quality vs performance tradeoff:

nearest neighbor (none)
bilinear (low)
bicubic (*)
mipmap (*)

The last 2 aren't mutually exclusive since you could apply nn/bilinear/bicubic even while doing mipmapping. There are also other filter algorighms, but we don't really embrace those. Also, mipmap is targeted at improving quality while reducing and just looks like bilinear (or bicubic potentially) while enlarging.

If I'd get rid of any of the names I'd probably get rid of "none" as it is just such a bad name. It implies that nothing will happen if taken out of context and it falls out of the spectrum of "low/med/hi". I might have named it "fastest" or "lowest" originally, ignoring whether 3 or 4 constants are useful for Flutter apps.

**Comment by yjbanov** [comment]
Yeah, so a more advanced API for picking a sampling matrix and/or mipmapping can be provided separately from the `FilterQuality` enum. `FilterQuality` could be the "non-advanced" API, with simple explanations and unsurprising behavior. Currently, it requires elaborate explanations (as you can see from this PR) and surprising behavior (e.g. `medium` is better quality than `high`, what?!). I think none (or low)/medium/high is good enough for the simple case. The sampling algorithm behind today's `high` does not qualify as "simple" and should be moved to the advanced API (which I wouldn't rush to add until users ask for it).

**Comment by flar** [comment]
To that end, Skia is moving to SkSamplingOptions, of which SkFilterQuality is only one component...

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -825,7 +825,24 @@ enum BlendMode {
...
+/// When scaling down, the higher the scale factor the less effective the [high]
+/// level is. Beyond certain scale level [medium] may provide better visual
+/// result than [high] (see example below).
+///
+/// When building for the web using the `--web-renderer=html` option, filter quality has
+/// no effect. All images are rendered using the respective browser's default setting.
+///
+/// Example:
+///
+/// ![](https://flutter.github.io/assets-for-api-docs/assets/dart-ui/filter_quality.png)
```
Comment (Hixie) [inline lib/ui/painting.dart:840]: We should offer some alternative text that can take the place of (be an alternative to) the image for when people don't have images enabled. (Or, alternatively, rephrase the text to not refer to the example, such that it reads naturally with or without the image.) See https://html.spec.whatwg.org/multipage/images.html#alt

Comment (yjbanov) [reply lib/ui/painting.dart:840]: I followed the other examples in our code and rephrased the text to not refer to the example.

**Comment by yjbanov** [comment]
I turned the discussion about actually changing the API into a separate feature request: https://github.com/flutter/flutter/issues/76737

**Comment by chinmaygarde** [comment]
Ping @Hixie for followup.

**Comment by flar** [comment]
See also https://github.com/flutter/engine/pull/24797

**Review by Hixie — APPROVED** [review]
I guess LGTM since accurate documentation is better than inaccurate documentation, but I agree that it seems the real fix here is fixing the implementation.

Can't we just make the implementation of "high" act like "medium" when shrinking, if that's really better?

**Comment by yjbanov** [comment]
> Can't we just make the implementation of "high" act like "medium" when shrinking, if that's really better?

This may be doable in Skia, but it's probably quite difficult to do on the Flutter side (engine or framework). The issue is that we don't have a way to tell if we're shrinking the image or magnifying it. It depends on the aggregate transform matrix that includes all layers as well as transforms inside the picture. Pictures are opaque and [as of today](https://github.com/flutter/flutter/issues/53501) backed by `SkPicture` directly. There's no way for us to intercept the rendering process and switch the sampling algorithm. I'm guessing Skia can probably do this, or perhaps provide an API for us to customize image rendering. If we go the custom recording format route we may not need anything extra from Skia (Android and Chrome went this route). However, if we want to keep the direct link to `SkPicture` we need more features from Skia.

**Comment by Hixie** [comment]
I meant in Skia, yes.

**Pull request merged**